# Runs of homozygosity 

This README describes the steps that need to be taken to run the runs of homozygosity pipeline as presented in Bortoluzzi et al. (2020).

If you are using any of these scripts, please cite the following papers:
* Bortoluzzi, C., Bosse, M., Derks, M. F., Crooijmans, R. P., Groenen, M. A., & Megens, H. J. (2020). The type of bottleneck matters: Insights into the deleterious variation landscape of small managed populations. Evolutionary applications, 13(2), 330-341.
* Bosse et al. (2012). Regions of homozygosity in the porcine genome: consequence of demography and the recombination landscape. PLoS genetics, 8(11).

## Description

This pipeline consists of various Bash and Python scripts, each described in the order they should be executed.

## Getting Started

### Dependencies

Before running the pipeline, make sure you have installed the following programs: 

* DeepVariant (v1.9.0): https://github.com/google/deepvariant?tab=readme-ov-file
* Samtools (v1.22.1): https://github.com/samtools/samtools
* Python >v3.6


### Installing

If you want to run this pipeline, you can simply clone the github repository using the code below:
```
git clone git@github.com:cbortoluzzi/ROHs.git
```

## Executing program

### Step 1: Call variants using DeepVariant
In this step, you will use DeepVariant to perform variant calling from sequencing reads: 

```deepvariant.sh <reference_genome> <bam> <model>```

Required inputs:
- Reference genome (```<reference_genome>```): the reference genome must be in uncompressed .fasta fomrat and indexed. Use the following command to index it:
```samtools faidx <reference_genome>```. 

- BAM file (```<bam>```): the input alignment must also be indexed:
```samtools index <bam>```.

- Model type (```<model>```): this specifies the DeepVariant model to use. Use ```WGS``` for Illumina whole-genome sequencing data, ```PACBIO```for PacBio data, or ```ONT_R104``` for Oxford Nanopore R10.4.1 chemistry Simplex and Duplex reads.

Outputs:
- VCF file: this is the standard VCF file containing all called variants.
- XML report: a report in XML format generated by DeepVariant containing summary statistics. For more information about the VCF stats report, refer to the DeepVariant documentation: https://github.com/google/deepvariant/blob/r1.9/docs/deepvariant-vcf-stats-report.md. 

**Important**: before running the script, make sure to update the paths to DeepVariant and the directory where your data is stored.

### Step 2: Calculate genome-wide coverage
In this step, you will use Samtools to compute the genome-wide coverage of each individual in your dataset: 

```samtools_depth.sh <bam>```

Required inputs:
- BAM file (```<bam>```): the indexed alignment file for the individual whose genome-wide coverage you want to calculate. 

Output:
- Coverage file: the bash script will produce a coverage file (```.cov```) reporting the genome-wide coverage. You will need this file in step 3 and 4.  

### Step 3: Filter called variants based on phred-quality score, read depth, and genotype quality
In this step, you will use a python script to filter variants called in step 1: 

```
python filter_variants.py -h

Parameters:

    --vcf    This is the VCF file generated by DeepVariant in step 1.

    --cov    This is the text file generated by Samtools in step 2.

    --q      This is the minimum Phred-quality score that a variant must have to be retained (default value: 15).

    --dp     This is the minimum read depth that a variant must have to be retained (default value: 6).

    --gq     This is the minimum genotype quality that a variant must have to be retained (default value: 20).

    --o      This is the name of the output directory where the filtered VCF file will be saved.
```


To run this script, make sure to have the following modules installed in python: 
- ```PyVCF 0.6.8```: to install it, refer to this page: https://pypi.org/project/PyVCF/
- ```pathlib 1.0.1```: to install it, refer to this page: https://pypi.org/project/pathlib/

This script will filter variants in the VCF file based on a minimum phred-quality score (default: 15), a minimum read depth (default: 6), and minimum genotype quality (default: 20). As these parameters can be changed when running the script (```-q```, ```-dp```, and ```gq``` option), I encourage you to have a look at the XML report generated in step 1 to check the quality of your data. It is possible that, for example, a genotype quality of 20 (default value) is too low or too high for your individual.

**Important**: the script fill also filter variants if their read depth is above 2 x the average genome-wide coverage. If you want to allow for a higher maximum read depth, you will need to manually change the script at line 33. 


## Help

Any advise for common problems or issues.
```
command to run if program contains helper info
```

## Authors

Contributors names and contact info

ex. Dominique Pizzie  
ex. [@DomPizzie](https://twitter.com/dompizzie)

## Version History

* 0.2
    * New release with documentation
* 0.1
    * Initial Release

## License

This project is licensed under the [NAME HERE] License - see the LICENSE.md file for details

## Acknowledgments

This pipeline was inspired by:
* Bosse et al. (2012). Regions of homozygosity in the porcine genome: consequence of demography and the recombination landscape. PLoS genetics, 8(11).

