# Runs of homozygosity 

This README describes the steps that need to be taken to run the runs of homozygosity pipeline as presented in Bortoluzzi et al. (2020).

If you are using any of these scripts, please cite the following papers:
* Bortoluzzi, C., Bosse, M., Derks, M. F., Crooijmans, R. P., Groenen, M. A., & Megens, H. J. (2020). The type of bottleneck matters: Insights into the deleterious variation landscape of small managed populations. Evolutionary applications, 13(2), 330-341.
* Bosse et al. (2012). Regions of homozygosity in the porcine genome: consequence of demography and the recombination landscape. PLoS genetics, 8(11).

## Description

This pipeline consists of various Bash and Python scripts, each described in the order they should be executed.

## Getting Started

### Dependencies

Before running the pipeline, make sure you have installed the following programs: 

* DeepVariant (v1.9.0): https://github.com/google/deepvariant?tab=readme-ov-file
* Samtools (v1.22.1): https://github.com/samtools/samtools
* Python >v3.6


### Installing

If you want to run this pipeline, you can simply clone the github repository using the code below:
```
git clone git@github.com:cbortoluzzi/ROHs.git
```

## Executing program

### Step 1: Call variants using DeepVariant
In this step, you will use DeepVariant to perform variant calling from sequencing reads: 

```deepvariant.sh <reference_genome> <bam> <model>```

Required inputs:
- Reference genome (```<reference_genome>```): the reference genome must be in uncompressed .fasta fomrat and indexed. Use the following command to index it:
```samtools faidx <reference_genome>```. 

- BAM file (```<bam>```): the input alignment must also be indexed:
```samtools index <bam>```.

- Model type (```<model>```): this specifies the DeepVariant model to use. Use ```WGS``` for Illumina whole-genome sequencing data, ```PACBIO```for PacBio data, or ```ONT_R104``` for Oxford Nanopore R10.4.1 chemistry Simplex and Duplex reads.

Outputs:
- VCF file: this is the standard VCF file containing all called variants.
- XML report: a report in XML format generated by DeepVariant containing summary statistics. For more information about the VCF stats report, refer to the DeepVariant documentation: https://github.com/google/deepvariant/blob/r1.9/docs/deepvariant-vcf-stats-report.md. 

**Important**: before running the script, make sure to update the paths to DeepVariant and the directory where your data is stored.

### Step 2: Calculate genome-wide coverage
In this step, you will use Samtools to compute the genome-wide coverage of each individual in your dataset: 

```samtools_depth.sh <bam>```

Required inputs:
- BAM file (```<bam>```): the indexed alignment file for the individual whose genome-wide coverage you want to calculate. 

Output:
- Coverage file: the bash script will produce a coverage file (```.cov```) reporting the genome-wide coverage. You will need this file in step 4, so for now keep it aside. 


## Help

Any advise for common problems or issues.
```
command to run if program contains helper info
```

## Authors

Contributors names and contact info

ex. Dominique Pizzie  
ex. [@DomPizzie](https://twitter.com/dompizzie)

## Version History

* 0.2
    * New release with documentation
* 0.1
    * Initial Release

## License

This project is licensed under the [NAME HERE] License - see the LICENSE.md file for details

## Acknowledgments

This pipeline was inspired by:
* Bosse et al. (2012). Regions of homozygosity in the porcine genome: consequence of demography and the recombination landscape. PLoS genetics, 8(11).

